From 4aa867593214a882d545b7432f36891932d05eae Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Mon, 5 Oct 2026 21:49:33 +0200
Subject: [PATCH 17/21] MFront: behaviour variables, models and porous flows
 usable in device code

- behaviour variables: default constructors of the behaviour and of its
  data classes and behaviour variable factories are host and device
  functions; in device code, parameters take their default values
- StandardElastoViscoPlasticity brick: device compatible failure of the
  staggered scheme of the porosity, yield radius defined by data usable
  in device code
- fix the name of the derivative of the yield radius interpolated by
  cubic splines when an identifier is used
- generic model interface: the class implementing a model is written by
  writeGenericModelClass, which can also generate a class usable in
  device code
- fix models with material properties and parameters treated as static
  variables, and static variables of models using quantities
---
 .../MFront/BehaviourCodeGeneratorBase.hxx     |   9 +
 .../include/MFront/GenericModelInterface.hxx  |  43 ++
 mfront/src/BehaviourCodeGeneratorBase.cxx     | 167 ++++---
 mfront/src/DataIsotropicHardeningRule.cxx     |  10 +-
 mfront/src/GenericModelInterface.cxx          | 452 +++++++++++-------
 .../StandardElastoViscoPlasticityBrick.cxx    |  13 +-
 6 files changed, 435 insertions(+), 259 deletions(-)

diff --git a/mfront/include/MFront/BehaviourCodeGeneratorBase.hxx b/mfront/include/MFront/BehaviourCodeGeneratorBase.hxx
index 8fe6c5c3a..4a5038be0 100644
--- a/mfront/include/MFront/BehaviourCodeGeneratorBase.hxx
+++ b/mfront/include/MFront/BehaviourCodeGeneratorBase.hxx
@@ -577,6 +577,15 @@ namespace mfront {
      */
     virtual void writeBehaviourParameterInitialisation(
         std::ostream&, const Hypothesis, const std::string&) const;
+    /*!
+     * \brief write the initialisation of the parameters to their default
+     * values in a constructor. The parameters treated as static variables
+     * are only initialised if they are declared in a base class.
+     * \param[in] os: output stream
+     * \param[in] h: modelling hypothesis
+     */
+    virtual void writeBehaviourParameterInitialisationToDefaultValues(
+        std::ostream&, const Hypothesis) const;
     /*!
      * \brief write code which differs on the host and in device code (see
      * the `TFEL_DEVICE_CODE` macro). This is mostly meant to replace
diff --git a/mfront/include/MFront/GenericModelInterface.hxx b/mfront/include/MFront/GenericModelInterface.hxx
index 2aea07190..70e067b5b 100644
--- a/mfront/include/MFront/GenericModelInterface.hxx
+++ b/mfront/include/MFront/GenericModelInterface.hxx
@@ -18,6 +18,49 @@
 
 namespace mfront {
 
+  /*!
+   * \brief options of the generation of the class implementing a model
+   */
+  struct GenericModelClassOptions {
+    /*!
+     * \brief if true, the class can be used in device code: its constructor
+     * takes the behaviour data and a structure holding the values of the
+     * real parameters, its methods return false if a bounds check fails
+     * and an `execute` method executes all the functions of the model.
+     */
+    bool device = false;
+    //! \brief type of the structure holding the values of the parameters
+    std::string parameters_type = "";
+  };
+
+  /*!
+   * \brief write the class implementing a model in the `mfront::gm`
+   * namespace, and the class holding the values of its real parameters (see
+   * `getGenericModelParametersInitializerName`)
+   * \param[out] os: output stream
+   * \param[in] fd: file description
+   * \param[in] md: model description
+   * \param[in] o: options
+   */
+  MFRONT_VISIBILITY_EXPORT void writeGenericModelClass(
+      std::ostream&,
+      const FileDescription&,
+      const ModelDescription&,
+      const GenericModelClassOptions&);
+  //! \return true if the given parameter is a real parameter
+  MFRONT_VISIBILITY_EXPORT bool isGenericModelRealParameter(
+      const VariableDescription&);
+  /*!
+   * \return true if the values of the real parameters of the given model are
+   * held by the class whose name is returned by
+   * `getGenericModelParametersInitializerName`
+   */
+  MFRONT_VISIBILITY_EXPORT bool hasGenericModelParametersInitializer(
+      const ModelDescription&);
+  //! \return the name of the class holding the values of the parameters
+  MFRONT_VISIBILITY_EXPORT std::string getGenericModelParametersInitializerName(
+      const ModelDescription&);
+
   /*!
    * \brief a generic interface for point-wise models implemented using the
    * `Model` domain specific language.
diff --git a/mfront/src/BehaviourCodeGeneratorBase.cxx b/mfront/src/BehaviourCodeGeneratorBase.cxx
index f1cbad41b..27c887dec 100644
--- a/mfront/src/BehaviourCodeGeneratorBase.cxx
+++ b/mfront/src/BehaviourCodeGeneratorBase.cxx
@@ -1330,7 +1330,7 @@ namespace mfront {
     os << "/*!\n"
        << "* \\brief Default constructor\n"
        << "*/\n"
-       << this->bd.getClassName() << "BehaviourData()\n"
+       << "TFEL_HOST_DEVICE " << this->bd.getClassName() << "BehaviourData()\n"
        << "{}\n\n"
        << "/*!\n"
        << "* \\brief copy constructor\n"
@@ -2432,25 +2432,44 @@ namespace mfront {
   void BehaviourCodeGeneratorBase::writeBehaviourConstructors(
       std::ostream& os, const Hypothesis h) const {
     auto tmpnames = std::vector<std::string>{};
-    auto write_body = [this, &os, h](const std::string& parameters_source) {
+    // \param[in] parameters_source: object giving the values of the
+    // parameters (see writeBehaviourParameterInitialisation)
+    // \param[in] device: if true, the parameters are initialised to their
+    // default values in device code, where the parameters initializers can't
+    // be used
+    auto write_body = [this, &os, h](const std::string& parameters_source,
+                                     const bool device = false) {
       os << "using namespace std;\n"
          << "using namespace tfel::math;\n"
          << "using std::vector;\n";
       writeMaterialLaws(os, this->bd.getMaterialLaws());
-      this->writeBehaviourParameterInitialisation(os, h, parameters_source);
+      const auto& parameters = this->bd.getBehaviourData(h).getParameters();
+      if ((device) && (parameters_source.empty()) &&
+          (!areParametersTreatedAsStaticVariables(this->bd)) &&
+          (!parameters.empty())) {
+        os << "#ifdef TFEL_DEVICE_CODE\n";
+        this->writeBehaviourParameterInitialisationToDefaultValues(os, h);
+        os << "#else /* TFEL_DEVICE_CODE */\n";
+        this->writeBehaviourParameterInitialisation(os, h, "");
+        os << "#endif /* TFEL_DEVICE_CODE */\n";
+      } else {
+        this->writeBehaviourParameterInitialisation(os, h, parameters_source);
+      }
     };
     this->checkBehaviourFile(os);
     // initializers
     const auto& init = this->getBehaviourConstructorsInitializers(h);
     // writing constructors
     if (this->bd.shallDefineDefaultConstructor()) {
+      // the default constructor is used by behaviour variables, which may be
+      // used in device code
       os << "//! \\brief default constructor\n"
-         << this->bd.getClassName() << "()";
+         << "TFEL_HOST_DEVICE " << this->bd.getClassName() << "()";
       if (!init.empty()) {
         os << "\n: " << init;
       }
       os << "\n{";
-      write_body("");
+      write_body("", true);
       os << "}\n\n";
     }
     os << "/*!\n"
@@ -3505,77 +3524,89 @@ namespace mfront {
     this->writeBehaviourParameterInitialisation(os, h, "");
   }  // end of writeBehaviourParameterInitialisation
 
-  void BehaviourCodeGeneratorBase::writeBehaviourParameterInitialisation(
-      std::ostream& os, const Hypothesis h, const std::string& src) const {
-    constexpr auto uh = ModellingHypothesis::UNDEFINEDHYPOTHESIS;
+  void BehaviourCodeGeneratorBase::
+      writeBehaviourParameterInitialisationToDefaultValues(
+          std::ostream& os, const Hypothesis h) const {
     const auto use_static_variables =
         areParametersTreatedAsStaticVariables(this->bd);
     this->checkBehaviourFile(os);
     const auto& d = this->bd.getBehaviourData(h);
     for (const auto& p : d.getParameters()) {
-      if (use_static_variables) {
-        if (!p.getAttribute<bool>(
-                VariableDescription::variableDeclaredInBaseClass, false)) {
-          continue;
+      // parameters treated as static variables are only initialised if
+      // they are declared in a base class
+      if ((use_static_variables) &&
+          (!p.getAttribute<bool>(
+              VariableDescription::variableDeclaredInBaseClass, false))) {
+        continue;
+      }
+      if (p.type == "int") {
+        os << "this->" << p.name << " = "
+           << this->bd.getIntegerParameterDefaultValue(h, p.name) << ";\n";
+      } else if (p.type == "ushort") {
+        os << "this->" << p.name << " = "
+           << this->bd.getUnsignedShortParameterDefaultValue(h, p.name)
+           << ";\n";
+      } else {
+        const auto f = SupportedTypes::getTypeFlag(p.type);
+        if (f != SupportedTypes::SCALAR) {
+          this->throwRuntimeError(
+              "BehaviourCodeGeneratorBase::"
+              "writeBehaviourParameterInitialisationToDefaultValues",
+              "unsupported parameter type '" + p.type +
+                  "' "
+                  "for parameter '" +
+                  p.name + "'");
         }
-        if (p.type == "int") {
-          os << "this->" << p.name << " = "
-             << this->bd.getIntegerParameterDefaultValue(h, p.name) << ";\n";
-        } else if (p.type == "ushort") {
-          os << "this->" << p.name << " = "
-             << this->bd.getUnsignedShortParameterDefaultValue(h, p.name)
-             << ";\n";
+        if (p.arraySize == 1u) {
+          os << "this->" << p.name << " = " << p.type << "("
+             << this->bd.getFloattingPointParameterDefaultValue(h, p.name)
+             << ");\n";
         } else {
-          const auto f = SupportedTypes::getTypeFlag(p.type);
-          if (f != SupportedTypes::SCALAR) {
-            this->throwRuntimeError(
-                "BehaviourCodeGeneratorBase::"
-                "writeBehaviourParameterInitialisation",
-                "unsupported parameter type '" + p.type +
-                    "' "
-                    "for parameter '" +
-                    p.name + "'");
-          }
-          if (p.arraySize == 1u) {
-            os << "this->" << p.name << " = " << p.type << "("
-               << this->bd.getFloattingPointParameterDefaultValue(h, p.name)
+          for (unsigned short i = 0; i != p.arraySize; ++i) {
+            os << "this->" << p.name << "[" << i << "] = " << p.type << "("
+               << this->bd.getFloattingPointParameterDefaultValue(h, p.name, i)
                << ");\n";
-          } else {
-            for (unsigned short i = 0; i != p.arraySize; ++i) {
-              os << "this->" << p.name << "[" << i << "] = " << p.type << "("
-                 << this->bd.getFloattingPointParameterDefaultValue(h, p.name,
-                                                                    i)
-                 << ");\n";
-            }
           }
         }
-      } else {
-        const auto getter = [this, h, &p, &src] {
-          if (!src.empty()) {
-            return src;
-          }
-          if ((h == uh) || (this->bd.hasParameter(uh, p.name))) {
-            return this->bd.getClassName() + "ParametersInitializer::get()";
-          }
-          return this->bd.getClassName() + ModellingHypothesis::toString(h) +
-                 "ParametersInitializer::get()";
-        }();
-        if (p.arraySize == 1u) {
-          os << "this->" << p.name << " = ";
-          if (this->bd.useQt()) {
-            os << p.type << "(" << getter << "." << p.name << ");\n";
-          } else {
-            os << getter << "." << p.name << ";\n";
-          }
+      }
+    }
+  }  // end of writeBehaviourParameterInitialisationToDefaultValues
+
+  void BehaviourCodeGeneratorBase::writeBehaviourParameterInitialisation(
+      std::ostream& os, const Hypothesis h, const std::string& src) const {
+    constexpr auto uh = ModellingHypothesis::UNDEFINEDHYPOTHESIS;
+    if (areParametersTreatedAsStaticVariables(this->bd)) {
+      this->writeBehaviourParameterInitialisationToDefaultValues(os, h);
+      return;
+    }
+    this->checkBehaviourFile(os);
+    const auto& d = this->bd.getBehaviourData(h);
+    for (const auto& p : d.getParameters()) {
+      const auto getter = [this, h, &p, &src] {
+        if (!src.empty()) {
+          return src;
+        }
+        if ((h == uh) || (this->bd.hasParameter(uh, p.name))) {
+          return this->bd.getClassName() + "ParametersInitializer::get()";
+        }
+        return this->bd.getClassName() + ModellingHypothesis::toString(h) +
+               "ParametersInitializer::get()";
+      }();
+      if (p.arraySize == 1u) {
+        os << "this->" << p.name << " = ";
+        if (this->bd.useQt()) {
+          os << p.type << "(" << getter << "." << p.name << ");\n";
         } else {
-          os << "tfel::fsalgo::transform<" << p.arraySize << ">::exe("  //
-             << getter << "." << p.name << ".begin(), "                 //
-             << "this->" << p.name << ".begin(), "
-             << "[](const auto& mfront_parameter_value){\n"
-             << "return " << p.type
-             << "{static_cast<NumericType>(mfront_parameter_value)};\n"
-             << "});\n";
+          os << getter << "." << p.name << ";\n";
         }
+      } else {
+        os << "tfel::fsalgo::transform<" << p.arraySize << ">::exe("  //
+           << getter << "." << p.name << ".begin(), "                 //
+           << "this->" << p.name << ".begin(), "
+           << "[](const auto& mfront_parameter_value){\n"
+           << "return " << p.type
+           << "{static_cast<NumericType>(mfront_parameter_value)};\n"
+           << "});\n";
       }
     }
   }  // end of writeBehaviourParameterInitialisation
@@ -3689,12 +3720,13 @@ namespace mfront {
       write_wrapper(b);
       os << "struct " << factory << " final\n"
          << "{\n"
-         << factory << "(const " << this->bd.getClassName()
+         << "TFEL_HOST_DEVICE " << factory << "(const "
+         << this->bd.getClassName()
          << "& mfront_behaviour_reference_argument)\n"
          << ": mfront_behaviour_reference("
          << "mfront_behaviour_reference_argument)\n"
          << "{}\n"
-         << "[[nodiscard]] " << wrapper << " make(){\n"
+         << "[[nodiscard]] TFEL_HOST_DEVICE " << wrapper << " make(){\n"
          << "return " << wrapper << "{};\n"
          << "} // end of make\n"
          << "private:\n"
@@ -5658,7 +5690,8 @@ namespace mfront {
     os << "/*!\n"
        << "* \\brief Default constructor\n"
        << "*/\n"
-       << this->bd.getClassName() << "IntegrationData()\n"
+       << "TFEL_HOST_DEVICE " << this->bd.getClassName()
+       << "IntegrationData()\n"
        << "{}\n\n"
        << "/*!\n"
        << "* \\brief Copy constructor\n"
diff --git a/mfront/src/DataIsotropicHardeningRule.cxx b/mfront/src/DataIsotropicHardeningRule.cxx
index ba73e6c7c..593feef65 100644
--- a/mfront/src/DataIsotropicHardeningRule.cxx
+++ b/mfront/src/DataIsotropicHardeningRule.cxx
@@ -119,7 +119,7 @@ namespace mfront::bbrick {
             .ordinates_name = "mfront_yield_surface_values",
             .ordinates_type = "stress"};
     const auto etype_value = this->etype ? "true" : "false";
-    auto m = "stress mfront_computeYieldRadius" + local_id +
+    auto m = "TFEL_HOST_DEVICE stress mfront_computeYieldRadius" + local_id +
              "(const strain mfront_arg_p) const{\n";
     if (this->values.size() == 1) {
       const auto& v = *(this->values.begin());
@@ -138,7 +138,8 @@ namespace mfront::bbrick {
       m += "mfront_arg_p);\n";
     }
     m += "}\n";
-    m += "std::pair<stress, stress> mfront_computeYieldRadiusAndDerivative" +
+    m += "TFEL_HOST_DEVICE std::pair<stress, stress> "
+         "mfront_computeYieldRadiusAndDerivative" +
          local_id + "(const strain mfront_arg_p) const {\n";
     if (this->values.size() == 1) {
       const auto& v = *(this->values.begin());
@@ -177,7 +178,8 @@ namespace mfront::bbrick {
     const auto etype_value = this->etype ? "true" : "false";
     const auto m1 = "mfront_computeYieldRadius" + local_id;
     const auto m2 = "mfront_computeYieldRadiusAndDerivative" + local_id;
-    auto m = "stress " + m1 + "(const strain mfront_arg_p) const{\n";
+    auto m = "TFEL_HOST_DEVICE stress " + m1 +
+             "(const strain mfront_arg_p) const{\n";
     if (this->values.size() == 1) {
       const auto& v = *(this->values.begin());
       std::ostringstream os;
@@ -195,7 +197,7 @@ namespace mfront::bbrick {
       m += "return {};\n";
     }
     m += "} // end of " + m1 + "\n\n";
-    m += "std::pair<stress, stress> " + m2 + local_id +
+    m += "TFEL_HOST_DEVICE std::pair<stress, stress> " + m2 +
          "(const strain mfront_arg_p) const {\n";
     if (this->values.size() == 1) {
       const auto& v = *(this->values.begin());
diff --git a/mfront/src/GenericModelInterface.cxx b/mfront/src/GenericModelInterface.cxx
index 57dd13322..bf4f7fdfe 100644
--- a/mfront/src/GenericModelInterface.cxx
+++ b/mfront/src/GenericModelInterface.cxx
@@ -77,189 +77,43 @@ namespace mfront {
     }
   }  // end of writeScalarStandardTypedefs
 
-  GenericModelInterface::GenericModelInterface() = default;
-
-  std::string GenericModelInterface::getLibraryName(
-      const ModelDescription& md) const {
-    if (md.library.empty()) {
-      if (!md.material.empty()) {
-        return md.material + "-generic";
-      } else {
-        return "Model";
-      }
-    }
-    return md.library + "-generic";
-  }  // end of getLibraryName
-
-  std::string GenericModelInterface::getName() {
-    return "generic";
-  }  // end of getName
-
-  void GenericModelInterface::setOptions(const DataMap& opts) {
-    if (!opts.empty()) {
-      tfel::raise("no options expected for interface '" + this->getName() +
-                  "'");
-    }
-  }  // end of setOptions
-
-  std::pair<bool, GenericModelInterface::tokens_iterator>
-  GenericModelInterface::treatKeyword(const std::string& k,
-                                      const std::vector<std::string>& i,
-                                      tokens_iterator current,
-                                      const tokens_iterator) {
-    tfel::raise_if(
-        (std::find(i.begin(), i.end(), "generic") != i.end()) ||
-            (std::find(i.begin(), i.end(), "Generic") != i.end()),
-        "GenericModelInterface::treatKeyword: unsupported keyword '" + k + "'");
-    return {false, current};
-  }  // end of treatKeyword
-
-  void GenericModelInterface::declareReservedNames(std::set<std::string>&) {
-  }  // end of declareReservedNames
-
-  void GenericModelInterface::writeOutputFiles(const FileDescription& fd,
-                                               const ModelDescription& md) {
-    this->writeHeaderFile(fd, md);
-    this->writeSourceFile(fd, md);
-  }  // end of writeOutputFiles
-
-  void GenericModelInterface::writeHeaderFile(const FileDescription& fd,
-                                              const ModelDescription& md) {
-    using tfel::material::ModellingHypothesis;
-    auto raise = [](const std::string& m) {
-      tfel::raise("GenericModelInterface::writeHeaderFile: " + m);
-    };
-    tfel::system::systemCall::mkdir("include/MFront");
-    tfel::system::systemCall::mkdir("include/MFront/GenericModel");
-    const auto name = md.library + md.className;
-    const auto header = name + "-generic.hxx";
-    std::ofstream os("include/MFront/GenericModel/" + header);
-    if (!os) {
-      raise("could not open file '" + header + "'");
-    }
-    os << "/*!\n"
-       << " * \\file   " << header << '\n'
-       << " * \\brief  This file declares the generic interface for the "
-       << md.className << " model\n"
-       << " * \\author " << fd.authorName << '\n'
-       << " * \\date   " << fd.date << '\n'
-       << " */\n\n";
-
-    const auto hg = getHeaderGuard(md);
-    os << "#ifndef " << hg << "\n"
-       << "#define " << hg << "\n\n"
-       << "#include\"TFEL/Config/TFELConfig.hxx\"\n"
-       << "#include\"MFront/GenericBehaviour/BehaviourData.h\"\n\n";
-
-    writeExportDirectives(os, false);
-
-    os << "#ifdef __cplusplus\n"
-       << "extern \"C\"{\n"
-       << "#endif /* __cplusplus */\n\n";
-
-    if ((hasRealParameters(md)) &&
-        (!areParametersTreatedAsStaticVariables(md))) {
-      os << "/*!\n"
-         << " * \\brief function allowing to modify the parameters of the "
-         << name << " model\n"
-         << " * \\param[in] n: name of the parameter\n"
-         << " * \\param[in] v: value of the parameter\n"
-         << " */\n"
-         << "MFRONT_SHAREDOBJ int\n"
-         << name << "_setParameter(const char *const, const double);\n\n";
-    }
-
-    for (const auto& h : ModellingHypothesis::getModellingHypotheses()) {
-      const auto f = name + "_" + ModellingHypothesis::toString(h);
-      os << "/*!\n"
-         << " * \\brief function implementing the " << name << " model for the "
-         << ModellingHypothesis::toString(h) << " modelling hypothesis\n"
-         << " * \\param[in,out] d: material data\n"
-         << " */\n"
-         << "MFRONT_SHAREDOBJ int " << f
-         << "(mfront_gb_BehaviourData* const);\n\n";
-    }
+  bool isGenericModelRealParameter(const VariableDescription& p) {
+    return (p.type == "double") || (p.isScalar());
+  }  // end of isGenericModelRealParameter
 
-    os << "#ifdef __cplusplus\n"
-       << "}\n"
-       << "#endif /* __cplusplus */\n\n"
-       << "#endif /* " << hg << " */\n";
+  bool hasGenericModelParametersInitializer(const ModelDescription& md) {
+    return hasRealParameters(md);
+  }  // end of hasGenericModelParametersInitializer
 
-  }  // end of writeHeaderFile
+  std::string getGenericModelParametersInitializerName(
+      const ModelDescription& md) {
+    return md.className + "RealParametersInitializer";
+  }  // end of getGenericModelParametersInitializerName
 
-  void GenericModelInterface::writeSourceFile(const FileDescription& fd,
-                                              const ModelDescription& md) {
-    using tfel::material::ModellingHypothesis;
-    constexpr auto uh = ModellingHypothesis::UNDEFINEDHYPOTHESIS;
+  void writeGenericModelClass(std::ostream& os,
+                              const FileDescription& fd,
+                              const ModelDescription& md,
+                              const GenericModelClassOptions& o) {
     auto raise = [](const std::string& m) {
-      tfel::raise("GenericModelInterface::writeSourceFile: " + m);
+      tfel::raise("writeGenericModelClass: " + m);
     };
     auto getVariablePosition = [&raise](const auto& variables, const auto& n) {
       const auto pv = std::find_if(
-          variables.begin(), variables.end(), [&n, &raise](const auto o) {
-            if (o.getTypeFlag() != SupportedTypes::SCALAR) {
-              raise("unsupported type for variable " + o.name);
+          variables.begin(), variables.end(), [&n, &raise](const auto v) {
+            if (v.getTypeFlag() != SupportedTypes::SCALAR) {
+              raise("unsupported type for variable " + v.name);
             }
-            return o.name == n;
+            return v.name == n;
           });
       if (pv == variables.end()) {
         raise("variable " + n + " not found");
       }
       return pv - variables.begin();
     };  // end of getVariablePosition
-    const auto name = md.library + md.className;
-    const auto header = name + "-generic.hxx";
-    const auto src = name + "-generic.cxx";
-    const auto cn = md.className + "RealParametersInitializer";
-    std::ofstream os("src/" + src);
-    os.precision(14);
-    if (!os) {
-      raise("could not open file '" + src + "'");
-    }
-    os << "/*!\n"
-       << " * \\file   " << src << '\n'
-       << " * \\brief  This file declares the generic interface for the "
-       << md.className << " model\n"
-       << " * \\author " << fd.authorName << '\n'
-       << " * \\date   " << fd.date << '\n'
-       << " */\n\n"
-       << "#include <cmath>\n"
-       << "#include <cstring>\n"
-       << "#include <sstream>\n"
-       << "#include \"TFEL/PhysicalConstants.hxx\"\n"
-       << "#include \"TFEL/Config/TFELTypes.hxx\"\n"
-       << "#include \"TFEL/Math/Array/View.hxx\"\n"
-       << "#include\"TFEL/Math/General/DerivativeType.hxx\"\n";
-    if (useQuantities(md)) {
-      os << "#include \"TFEL/Math/qt.hxx\"\n";
-    }
-    os << "#include \"TFEL/Material/BoundsCheck.hxx\"\n";
-    os << "#include \"MFront/GenericBehaviour/Integrate.hxx\"\n\n"
-       << "#include \"MFront/GenericModel/" << header << "\"\n\n";
-    //
-    if (!md.includes.empty()) {
-      os << md.includes << "\n\n";
-    }
-    //
-    const auto mhs = ModellingHypothesis::getModellingHypotheses();
-    const auto hypotheses =
-        std::set<ModellingHypothesis::Hypothesis>{mhs.begin(), mhs.end()};
-    const auto bd = convertToBehaviourDescription(md);
-    GenericBehaviourSymbolsGenerator sg;
-    GenericBehaviourInterface bi;
-    sg.generateGeneralSymbols(os, bi, bd, fd, hypotheses, name);
-    os << '\n';
-    sg.generateSymbols(os, bi, bd, fd, name, uh);
-    //
-    if (!md.sources.empty()) {
-      os << md.sources << "\n\n";
-    }
-    //
+    const auto cn = getGenericModelParametersInitializerName(md);
     const auto has_constructor =
         (!md.constantMaterialProperties.empty()) || (hasRealParameters(md));
     //
-    os << "namespace mfront::gm{\n\n";
-    //
     if ((hasRealParameters(md)) &&
         (!areParametersTreatedAsStaticVariables(md))) {
       os << "struct " << cn << "{\n";
@@ -287,12 +141,24 @@ namespace mfront {
     os << "struct " << md.className << "{\n\n";
     writeScalarStandardTypedefs(os, md);
     os << '\n';
-    if (has_constructor) {
+    if (o.device) {
+      // the constructor always takes the behaviour data and the parameters
+      os << "TFEL_HOST_DEVICE " << md.className << "("
+         << "[[maybe_unused]] const mfront_gb_BehaviourData& "
+         << "mfront_model_data, "
+         << "[[maybe_unused]] const " << o.parameters_type
+         << "& mfront_parameters)";
+      if (has_constructor) {
+        os << "\n:";
+      }
+    } else if (has_constructor) {
       os << md.className << "(";
       if (!md.constantMaterialProperties.empty()) {
         os << "const mfront_gb_BehaviourData& mfront_model_data";
       }
       os << ")\n:";
+    }
+    if (has_constructor) {
       auto first = true;
       for (const auto& mp : md.constantMaterialProperties) {
         os << (first ? "" : ",\n");
@@ -302,22 +168,38 @@ namespace mfront {
         first = false;
       }
       for (const auto& p : md.parameters) {
-        if (!isRealParameter(p)) {
+        // parameters treated as static variables are not initialized here
+        if ((!isRealParameter(p)) || (!hasRealParameters(md))) {
           continue;
         }
         os << (first ? "" : ",\n");
-        os << p.name << "(" << cn << "::get()." << p.name << ")";
+        if (o.device) {
+          os << p.name << "(mfront_parameters." << p.name << ")";
+        } else {
+          os << p.name << "(" << cn << "::get()." << p.name << ")";
+        }
         first = false;
       }
-      os << "\n"
-         << "{}\n\n";
+      os << "\n";
+    }
+    if ((o.device) || (has_constructor)) {
+      os << "{}\n\n";
     }
     for (const auto& f : md.functions) {
-      os << "void execute_" << f.name
-         << "(mfront_gb_BehaviourData& mfront_model_data) const{\n";
+      if (o.device) {
+        os << "TFEL_HOST_DEVICE bool execute_" << f.name
+           << "(mfront_gb_BehaviourData& mfront_model_data) const{\n";
+      } else {
+        os << "void execute_" << f.name
+           << "(mfront_gb_BehaviourData& mfront_model_data) const{\n";
+      }
       os << "using namespace std;\n"
          << "using namespace tfel::math;\n"
          << "using namespace tfel::material;\n";
+      if (o.device) {
+        // material laws callable in device code
+        writeMaterialLaws(os, md.materialLaws);
+      }
       if (useQuantities(md)) {
         os << "using PhysicalConstants [[maybe_unused]] = "
            << "tfel::PhysicalConstants<NumericType, true>;\n";
@@ -362,9 +244,13 @@ namespace mfront {
              << "]};\n";
         }
       }
-      auto write_physical_bounds = [&os, &raise](const VariableDescription& v,
-                                                 const std::string& n,
-                                                 const unsigned short vdepth) {
+      // in device code, a failure of a bounds check is reported by the
+      // returned value
+      const auto on_failure = o.device ? "return false;" : "";
+      auto write_physical_bounds = [&os, &raise, &on_failure](
+                                       const VariableDescription& v,
+                                       const std::string& n,
+                                       const unsigned short vdepth) {
         if (!v.hasPhysicalBounds()) {
           return;
         }
@@ -372,9 +258,10 @@ namespace mfront {
           raise("error while treating physical bounds of variable '" + v.name +
                 "' (only scalar variable are supported)");
         }
-        writePhysicalBoundsChecks(os, v, n, "1u", false, false);
+        writePhysicalBoundsChecks(os, v, n, "1u", false, false, on_failure);
         if (vdepth == 1u) {
-          writePhysicalBoundsChecks(os, v, v.name + "_1", "1u", false, false);
+          writePhysicalBoundsChecks(os, v, v.name + "_1", "1u", false, false,
+                                    on_failure);
         }
       };
       if (!areRuntimeChecksDisabled(md)) {
@@ -394,9 +281,10 @@ namespace mfront {
         }
       }
       //
-      auto write_bounds = [&os, &raise](const VariableDescription& v,
-                                        const std::string& n,
-                                        const unsigned short vdepth) {
+      auto write_bounds = [&os, &raise, &on_failure](
+                              const VariableDescription& v,
+                              const std::string& n,
+                              const unsigned short vdepth) {
         if (!v.hasBounds()) {
           return;
         }
@@ -404,9 +292,10 @@ namespace mfront {
           raise("error while treating bounds of variable '" + v.name +
                 "' (only scalar variable are supported)");
         }
-        writeBoundsChecks(os, v, n, "1u", "policy", false, false);
+        writeBoundsChecks(os, v, n, "1u", "policy", false, false, on_failure);
         if (vdepth == 1u) {
-          writeBoundsChecks(os, v, v.name + "_1", "1u", "policy", false, false);
+          writeBoundsChecks(os, v, v.name + "_1", "1u", "policy", false, false,
+                            on_failure);
         }
       };
       if (!areRuntimeChecksDisabled(md)) {
@@ -436,8 +325,28 @@ namespace mfront {
            << "]) = " << v.name << ";\n";
         static_cast<void>(vdepth);
       }
+      if (o.device) {
+        os << "return true;\n";
+      }
       os << "} // end of execute_" << f.name << "\n\n";
     }
+    if (o.device) {
+      os << "//! \\brief execute all the functions of the model\n"
+         << "TFEL_HOST_DEVICE bool "
+         << "execute(mfront_gb_BehaviourData& mfront_model_data) const{\n"
+         << "return ";
+      if (md.functions.empty()) {
+        os << "true";
+      }
+      auto first_function = true;
+      for (const auto& f : md.functions) {
+        os << (first_function ? "" : " &&\n") << "this->execute_" << f.name
+           << "(mfront_model_data)";
+        first_function = false;
+      }
+      os << ";\n"
+         << "} // end of execute\n\n";
+    }
     os << "private:\n";
     if (!md.members.empty()) {
       os << md.members << "\n\n";
@@ -462,6 +371,9 @@ namespace mfront {
         os << "const " << p.type << " " << p.name << ";\n";
         continue;
       }
+      if ((o.device) && (p.type == "string")) {
+        raise("string parameters are not supported in device code");
+      }
       if (p.type == "string") {
         os << "const std::string " << p.name << " = "
            << p.getAttribute<std::string>(VariableDescription::defaultValue);
@@ -481,10 +393,180 @@ namespace mfront {
       if (!getDebugMode()) {
         printLinePragma(os, v.lineNumber, fd.fileName);
       }
-      os << "static constexpr " << v.type << " " << v.name << " = "  //
-         << v.value << ";\n";
+      // brace initialisation is required by quantities
+      os << "static constexpr auto " << v.name << " = " << v.type << "{"
+         << v.value << "};\n";
     }
     os << "}; // end of struct " << md.className << "\n\n";
+  }  // end of writeGenericModelClass
+
+  GenericModelInterface::GenericModelInterface() = default;
+
+  std::string GenericModelInterface::getLibraryName(
+      const ModelDescription& md) const {
+    if (md.library.empty()) {
+      if (!md.material.empty()) {
+        return md.material + "-generic";
+      } else {
+        return "Model";
+      }
+    }
+    return md.library + "-generic";
+  }  // end of getLibraryName
+
+  std::string GenericModelInterface::getName() {
+    return "generic";
+  }  // end of getName
+
+  void GenericModelInterface::setOptions(const DataMap& opts) {
+    if (!opts.empty()) {
+      tfel::raise("no options expected for interface '" + this->getName() +
+                  "'");
+    }
+  }  // end of setOptions
+
+  std::pair<bool, GenericModelInterface::tokens_iterator>
+  GenericModelInterface::treatKeyword(const std::string& k,
+                                      const std::vector<std::string>& i,
+                                      tokens_iterator current,
+                                      const tokens_iterator) {
+    tfel::raise_if(
+        (std::find(i.begin(), i.end(), "generic") != i.end()) ||
+            (std::find(i.begin(), i.end(), "Generic") != i.end()),
+        "GenericModelInterface::treatKeyword: unsupported keyword '" + k + "'");
+    return {false, current};
+  }  // end of treatKeyword
+
+  void GenericModelInterface::declareReservedNames(std::set<std::string>&) {
+  }  // end of declareReservedNames
+
+  void GenericModelInterface::writeOutputFiles(const FileDescription& fd,
+                                               const ModelDescription& md) {
+    this->writeHeaderFile(fd, md);
+    this->writeSourceFile(fd, md);
+  }  // end of writeOutputFiles
+
+  void GenericModelInterface::writeHeaderFile(const FileDescription& fd,
+                                              const ModelDescription& md) {
+    using tfel::material::ModellingHypothesis;
+    auto raise = [](const std::string& m) {
+      tfel::raise("GenericModelInterface::writeHeaderFile: " + m);
+    };
+    tfel::system::systemCall::mkdir("include/MFront");
+    tfel::system::systemCall::mkdir("include/MFront/GenericModel");
+    const auto name = md.library + md.className;
+    const auto header = name + "-generic.hxx";
+    std::ofstream os("include/MFront/GenericModel/" + header);
+    if (!os) {
+      raise("could not open file '" + header + "'");
+    }
+    os << "/*!\n"
+       << " * \\file   " << header << '\n'
+       << " * \\brief  This file declares the generic interface for the "
+       << md.className << " model\n"
+       << " * \\author " << fd.authorName << '\n'
+       << " * \\date   " << fd.date << '\n'
+       << " */\n\n";
+
+    const auto hg = getHeaderGuard(md);
+    os << "#ifndef " << hg << "\n"
+       << "#define " << hg << "\n\n"
+       << "#include\"TFEL/Config/TFELConfig.hxx\"\n"
+       << "#include\"MFront/GenericBehaviour/BehaviourData.h\"\n\n";
+
+    writeExportDirectives(os, false);
+
+    os << "#ifdef __cplusplus\n"
+       << "extern \"C\"{\n"
+       << "#endif /* __cplusplus */\n\n";
+
+    if ((hasRealParameters(md)) &&
+        (!areParametersTreatedAsStaticVariables(md))) {
+      os << "/*!\n"
+         << " * \\brief function allowing to modify the parameters of the "
+         << name << " model\n"
+         << " * \\param[in] n: name of the parameter\n"
+         << " * \\param[in] v: value of the parameter\n"
+         << " */\n"
+         << "MFRONT_SHAREDOBJ int\n"
+         << name << "_setParameter(const char *const, const double);\n\n";
+    }
+
+    for (const auto& h : ModellingHypothesis::getModellingHypotheses()) {
+      const auto f = name + "_" + ModellingHypothesis::toString(h);
+      os << "/*!\n"
+         << " * \\brief function implementing the " << name << " model for the "
+         << ModellingHypothesis::toString(h) << " modelling hypothesis\n"
+         << " * \\param[in,out] d: material data\n"
+         << " */\n"
+         << "MFRONT_SHAREDOBJ int " << f
+         << "(mfront_gb_BehaviourData* const);\n\n";
+    }
+
+    os << "#ifdef __cplusplus\n"
+       << "}\n"
+       << "#endif /* __cplusplus */\n\n"
+       << "#endif /* " << hg << " */\n";
+
+  }  // end of writeHeaderFile
+
+  void GenericModelInterface::writeSourceFile(const FileDescription& fd,
+                                              const ModelDescription& md) {
+    using tfel::material::ModellingHypothesis;
+    constexpr auto uh = ModellingHypothesis::UNDEFINEDHYPOTHESIS;
+    auto raise = [](const std::string& m) {
+      tfel::raise("GenericModelInterface::writeSourceFile: " + m);
+    };
+    const auto name = md.library + md.className;
+    const auto header = name + "-generic.hxx";
+    const auto src = name + "-generic.cxx";
+    const auto cn = getGenericModelParametersInitializerName(md);
+    std::ofstream os("src/" + src);
+    os.precision(14);
+    if (!os) {
+      raise("could not open file '" + src + "'");
+    }
+    os << "/*!\n"
+       << " * \\file   " << src << '\n'
+       << " * \\brief  This file declares the generic interface for the "
+       << md.className << " model\n"
+       << " * \\author " << fd.authorName << '\n'
+       << " * \\date   " << fd.date << '\n'
+       << " */\n\n"
+       << "#include <cmath>\n"
+       << "#include <cstring>\n"
+       << "#include <sstream>\n"
+       << "#include \"TFEL/PhysicalConstants.hxx\"\n"
+       << "#include \"TFEL/Config/TFELTypes.hxx\"\n"
+       << "#include \"TFEL/Math/Array/View.hxx\"\n"
+       << "#include\"TFEL/Math/General/DerivativeType.hxx\"\n";
+    if (useQuantities(md)) {
+      os << "#include \"TFEL/Math/qt.hxx\"\n";
+    }
+    os << "#include \"TFEL/Material/BoundsCheck.hxx\"\n";
+    os << "#include \"MFront/GenericBehaviour/Integrate.hxx\"\n\n"
+       << "#include \"MFront/GenericModel/" << header << "\"\n\n";
+    //
+    if (!md.includes.empty()) {
+      os << md.includes << "\n\n";
+    }
+    //
+    const auto mhs = ModellingHypothesis::getModellingHypotheses();
+    const auto hypotheses =
+        std::set<ModellingHypothesis::Hypothesis>{mhs.begin(), mhs.end()};
+    const auto bd = convertToBehaviourDescription(md);
+    GenericBehaviourSymbolsGenerator sg;
+    GenericBehaviourInterface bi;
+    sg.generateGeneralSymbols(os, bi, bd, fd, hypotheses, name);
+    os << '\n';
+    sg.generateSymbols(os, bi, bd, fd, name, uh);
+    //
+    if (!md.sources.empty()) {
+      os << md.sources << "\n\n";
+    }
+    //
+    os << "namespace mfront::gm{\n\n";
+    writeGenericModelClass(os, fd, md, GenericModelClassOptions{});
     //
 
     os << "static int " << name
diff --git a/mfront/src/StandardElastoViscoPlasticityBrick.cxx b/mfront/src/StandardElastoViscoPlasticityBrick.cxx
index 963899ee7..212f7a5a6 100644
--- a/mfront/src/StandardElastoViscoPlasticityBrick.cxx
+++ b/mfront/src/StandardElastoViscoPlasticityBrick.cxx
@@ -16,6 +16,7 @@
 #include "TFEL/Glossary/GlossaryEntry.hxx"
 #include "TFEL/Utilities/Data.hxx"
 #include "MFront/MFrontDebugMode.hxx"
+#include "MFront/CodeGeneratorUtilities.hxx"
 #include "MFront/ImplicitDSLBase.hxx"
 #include "MFront/AbstractNonLinearSystemSolver.hxx"
 #include "MFront/BehaviourBrick/BrickUtilities.hxx"
@@ -960,11 +961,17 @@ namespace mfront {
          << "} else {\n";
       generate_debug_message(false);
       // the staggered scheme has not converged
+      // in device code, the resolution is aborted by setting the iteration
+      // counter of the non linear solver to its maximum value
       os << "if(this->" << staggeredSchemeIterationCounter << " > this->"
          << maximumNumberOfIterationsOfTheStaggeredScheme << "){\n"
-         << "tfel::raise<DivergenceException>(\""
-         << "maximum number of iterations of "
-         << "the staggered algorithm reached\");\n"
+         << makeHostAndDeviceCode(
+                "tfel::raise<DivergenceException>(\""
+                "maximum number of iterations of "
+                "the staggered algorithm reached\");",
+                "converged = false;\n"
+                "this->iter = this->iterMax;\n"
+                "return;")
          << "}\n";
       // update the current estimate
       os << staggeredSchemeBissectionAlgorithm << ".updateBounds(this->"
-- 
2.43.0

